### PR TITLE
Feature/join redux

### PR DESCRIPTION
**Pull request by YauhenKoipish**
Add redux to project.

### CODE REVIEW

**Review by Larkgronick — DISMISSED** [review]

File: src/components/Forms/MemberForm/MemberForm.jsx [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
+import { connect } from 'react-redux';
 import { Component } from 'react';
 import { Button } from '../../Buttons/Button/Button';
 import { registerUser, setUserData, setUserValidation } from '../../../services/services';
 import { createFullName } from '../../../services/createFullName';
 import SelectControl from '../components/SelectControl/SelectControl';
 import { memberFormRolesList, memberFormDirectionsList } from '../../../services/fields-template';
 
-export class MemberForm extends Component {
+class MemberForm extends Component {
```
Comment (maximsan) [inline src/components/Forms/MemberForm/MemberForm.jsx:11]: use PureComponent whenever it's possible, read about it https://ru.reactjs.org/docs/react-api.html#reactpurecomponent

File: src/components/Forms/MemberForm/MemberForm.jsx [hunk 2]
```diff
@@ -296,11 +298,38 @@ export class MemberForm extends Component {
...
+  }),
   isReadOnly: PropTypes.bool.isRequired,
   isEditMode: PropTypes.bool.isRequired,
 };
 
 MemberForm.defaultProps = {
   memberData: null,
 };
+
+function mapStateToProps(state) {
```
Comment (maximsan) [inline src/components/Forms/MemberForm/MemberForm.jsx:327]: spread state

File: src/components/Header/Header.jsx [hunk 3]
```diff
@@ -1,25 +1,28 @@
...
-      <span className='header__content__text'> Dims-11</span>
-      <div className='header__content__user'>
-        <span className='header__content__user__name'>{userData.name}</span>
-        <Button className='header__content__user__btn' onClick={hadleClick}>
-          Logout
-        </Button>
+export const Header = () => {
+  const {
+    userData: { name, lastName },
+  } = useSelector((state) => state.app);
```
Comment (maximsan) [inline src/components/Header/Header.jsx:13]: we don't use hooks in this week, don't mix the approaches, use `connect`

File: src/components/SideBar/SideBar.jsx [hunk 4]
```diff
@@ -10,19 +11,24 @@ export const SideBar = () => {
     history.goBack();
   }
 
+  const {
+    userData: { role },
+  } = useSelector((state) => state.app);
```
Comment (maximsan) [inline src/components/SideBar/SideBar.jsx:16]: we don't use hooks in this week, don't mix the approaches, use ```connect```

File: src/components/SideBar/SideBar.jsx [hunk 5]
```diff
@@ -10,19 +11,24 @@ export const SideBar = () => {
...
-            <img alt='img' className='sideBar__container__elem__img' src={img} />
-            <span className='sideBar__container__elem__desc'> {name}</span>
-          </Link>
-        ))}
+        {role !== 'MEMBER' &&
+          sideBarList.map(({ path, name, img }) => (
+            <Link
+              to={path}
+              key={name}
+              className={`sideBar__container__elem ${document.location.pathname === path && 'active'}`}
```
Comment (maximsan) [inline src/components/SideBar/SideBar.jsx:26]: ```pathname ``` should be taken from react router dom props

File: src/store/actions/app.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+export function getUserData() {
+  return async (dispatch) => {
+    await firebase.auth().onAuthStateChanged(async (user) => {
+      if (user) {
+        const promise = Promise.all([
+          getUserById(user.uid),
+          getUserValidation(user.uid),
+          dispatch(getSnapshot('users', 'membersData')),
+          dispatch(getSnapshot('tasks', 'tasksData')),
```
Comment (maximsan) [inline src/store/actions/app.js:13]: don't need to call dispatches in ```Promise.all```, it doesn't make sense

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left several comments. Check them out, thanks!

File: src/components/SideBar/SideBar.jsx [hunk 7]
```diff
@@ -10,19 +11,24 @@ export const SideBar = () => {
...
-          <Link
-            to={path}
-            key={name}
-            className={`sideBar__container__elem ${document.location.pathname === path && 'active'}`}
-          >
-            <img alt='img' className='sideBar__container__elem__img' src={img} />
-            <span className='sideBar__container__elem__desc'> {name}</span>
-          </Link>
-        ))}
+        {role !== 'MEMBER' &&
```
Comment (artur-nikitsin) [inline src/components/SideBar/SideBar.jsx:21]: You can store string values in separate object to avoid errors

**Review by artur-nikitsin — COMMENTED** [review]
Hi! Nice Work!

File: src/App/App.jsx [hunk 8]
```diff
@@ -1,113 +1,48 @@
...
           condition={isAuth && !isVerified}
-          userData={userData}
-          tasksData={tasksData}
           component={PasswordChange}
           path='/password-change'
           redirectPath='/members'
           exact
         />
-        {isAuth ? <Redirect to='/password-change' /> : <Route path='/' render={(props) => <Login />} />}
+        {isAuth ? <Redirect to='/password-change' /> : <Route path='/' render={() => <Login />} />}
```
[the remark this answers is not in the document]
Comment (YauhenKoipish) [reply src/App/App.jsx:45]: thx, i fixed it

File: src/components/Forms/MemberForm/MemberForm.jsx (hunk 2 above)
Comment (YauhenKoipish) [reply src/components/Forms/MemberForm/MemberForm.jsx:327]: Thanks, I changed this, and in other files too.
```suggestion
function mapStateToProps({ members: { memberData, isEditMode, isReadOnly } }) {
```

File: src/store/actions/app.js (hunk 6 above)
Comment (YauhenKoipish) [reply src/store/actions/app.js:13]: thx, i fixed this.

File: src/App/App.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (YauhenKoipish) [reply src/App/App.jsx:45]: thx, i fixed it.
```suggestion
        {isAuth ? <Redirect to='/password-change' /> : <Route path='/' component={Login} />}
```

File: src/components/SideBar/SideBar.jsx (hunk 5 above)
Comment (YauhenKoipish) [reply src/components/SideBar/SideBar.jsx:26]: thx, i fixed it.

File: src/components/SideBar/SideBar.jsx (hunk 7 above)
Comment (YauhenKoipish) [reply src/components/SideBar/SideBar.jsx:21]: thx for it, i changed it.

File: src/components/Header/Header.jsx (hunk 3 above)
Comment (YauhenKoipish) [reply src/components/Header/Header.jsx:13]: Thx, i fixed it.

File: src/components/SideBar/SideBar.jsx (hunk 4 above)
Comment (YauhenKoipish) [reply src/components/SideBar/SideBar.jsx:16]: Thx, i fixed it.

File: src/components/Forms/MemberForm/MemberForm.jsx (hunk 1 above)
Comment (YauhenKoipish) [reply src/components/Forms/MemberForm/MemberForm.jsx:11]: thx, I fixed it, I will keep this in mind in the future.

**Review by maximsan — APPROVED** [review]
Looks ok. I will proceed with merge, but need more approves

File: src/App/App.jsx [hunk 9]
```diff
@@ -1,150 +1,106 @@
...
-import { Members } from '../pages/Members';
-import { MemberProgress } from '../pages/MemberProgress';
-import { MemberTasks } from '../pages/MemberTasks';
-import { Tasks } from '../pages/Tasks';
-import { TaskTracks } from '../pages/TaskTracks';
+import Members from '../pages/Members';
+import MemberProgress from '../pages/MemberProgress';
+import MemberTasks from '../pages/MemberTasks';
+import Tasks from '../pages/Tasks';
+import TaskTracks from '../pages/TaskTracks';
```
Comment (Larkgronick) [inline src/App/App.jsx:11]: You can use multiimport here. It looks like that:
```
//App.jsx
import { Members, MemberProgress , MemberTasks, Tasks ,  TaskTracks  } from './index';
```
Before that you need to create separate index.js file, in wich you can place all components:
```
//index.js
export { Members } from '../pages/Members';
export { MemberProgress } from '../pages/MemberProgress ';
....
```

**Review by Larkgronick — APPROVED** [review]

**Pull request merged**